plugins/inputs/ctrlx_datalayer_sse/ctrlx_datalayer_sse.go

```diff
@@ -0,0 +1,386 @@
+			return fmt.Errorf("subscription %d: setting 'queue_behaviour' %q is invalid", i, sub.QueueBehaviour)
+		}
+		err = choice.Check(sub.ValueChange, valueChanges)
+		if err != nil {
+			s.Log.Infof("The right value change values are %v", valueChanges)
+			return fmt.Errorf("subscription %d: setting 'value_change' %q is invalid", i, sub.ValueChange)
+		}
+		if len(sub.Nodes) == 0 {
+			s.Log.Error("A configured subscription has no nodes configured")
+		}
+		sub.index = i
+	}
+
+	// Generate valid communication url based on configured hostname
+	u := url.URL{
+		Scheme: "https",
+		Host:   s.Hostname,
+	}
+	s.url = u.String()
+	_, err := url.Parse(s.url)
+	if err != nil {
+		return errors.New("invalid hostname")
+	}
+
+	return nil
+}
+
+// Start input as service, retain the accumulator, establish the connection.
+func (s *CtrlXDataLayerSSE) Start(acc telegraf.Accumulator) error {
+	var ctx context.Context
+	ctx, s.cancel = context.WithCancel(context.Background())
+
+	h := &httpconfig.HTTPClientConfig{
+		Timeout: s.Timeout,
+		ClientConfig: tls.ClientConfig{
+			InsecureSkipVerify: s.InsecureSkipVerify,
+		},
+	}
+
+	var err error
+	s.connection, err = h.CreateClient(ctx, s.Log)
+	if err != nil {
+		return fmt.Errorf("failed to create http client, err: %w", err)
+	}
+
+	s.tokenManager = token.TokenManager{
+		Url:        s.url,
+		Username:   s.Username,
+		Password:   s.Password,
+		Connection: s.connection,
+	}
+
+	s.acc = acc
+
+	s.gatherLoop(ctx)
+
+	return nil
+}
+
+// gatherLoop creates sse subscriptions on the Data Layer and requests the sse data
+// the connection will be restablished if the sse subscription is broken.
+func (s *CtrlXDataLayerSSE) gatherLoop(ctx context.Context) {
+	for _, sub := range s.Subscription {
+		s.wg.Add(1)
+		go func(sub Subscription) {
+			defer s.wg.Done()
+			for {
+				select {
+				case <-ctx.Done():
+					s.Log.Debugf("Gather loop for subscription %d stopped", sub.index)
+					return
+				default:
+					client, err := s.createSubscriptionAndSseClient(&sub)
+					if err != nil {
+						s.Log.Errorf("Creating sse client to subscription %d, %v", sub.index, err)
+						time.Sleep(time.Duration(defaultReconnectInterval))
+						continue
+					}
+					s.Log.Debugf("Created sse client to subscription %d", sub.index)
+
+					// Read the server-sent events and add all received data to
+					// the accumulator. This function will only terminate in case
+					// of connection errors (err != nil) or if the subscription is terminated
+					// by the telegraf side (err == nil).
+					err = client.Subscribe(ctx, func(event string, data string) {
+						s.addMetric(&sseclient.SseEvent{
+							Event: event,
+							Data:  data,
+						}, &sub)
+					})
+					if err == nil {
+						// Subscription cancelled
+						s.Log.Debugf("Requesting data of subscription %d cancelled", sub.index)
+						return
+					}
+					s.Log.Errorf("Requesting data of subscription %d failed, %v", sub.index, err)
```

Comment:
I don't think the comment matches the code. Especially the

> This function will only terminate in case of connection errors (err != nil) [...]

part is not obvious to me.